Lib/python/pyprimtypes.swg

```diff
@@ -104,7 +104,9 @@ SWIG_AsVal_dec(long)(PyObject *obj, long* val)
     if (!dispatch) {
       double d;
       int res = SWIG_AddCast(SWIG_AsVal(double)(obj,&d));
-      if (SWIG_IsOK(res) && SWIG_CanCastAsInteger(&d, LONG_MIN, LONG_MAX)) {
+      // largest double not larger than LONG_MAX
+      const double long_max = sizeof(long) == 8 ? 0x7ffffffffffffc00LL : LONG_MAX;
```

Comment:
I really hate hardcoding this two constants. Howver I was not able to generate this constant in C with less than 10 lines of code without any a-prioi-knowledge. I anyone is aware of a method to convert a 64 bit integer > 2**53 to double with a given rounding behavior, I will happily update this patch.

How to generate:
  C: nextafter(uint128_t(LONG_MAX)+1, 0)
  Python: hex(int(nextafter(2**63, 0)))

Reply:
This also seems to assume IEEE double, which is pretty much ubiquitous in current platforms but not actually required by the language standards.

It does seem fiddly to calculate though.

Reply:
If we have `uint128_t` (or if `sizeof(long) == 4` and we have `uint64_t`) and we're building as C99 or have POSIX functions then we could just use `nextafter((uint128_t)LONG_MAX+1, 0)` (or `nextafter((uint64_t)LONG_MAX+1, 0)`) - checking on godbolt it seems GCC and MSVC both evaluate this at compile time (though clang doesn't seem to currently).  We could also check `FLT_RADIX` and `DBL_MANT_DIG` match IEEE and use a precomputed value for common platforms too.

Many platforms lack `uint128_t` though, and I'd guess a platform not using IEEE floating point is probably especially unlikely to.

Some other assumptions would be easy to fix though, e.g.:

```
llmaxd = float(2**53) # 64 bit double significand
```

We could add C helpers to return `FLT_RADIX` and `DBL_MANT_DIG` (from `<float.h>`) and replace this with something like:

```
llmaxd = float(flt_radix()**dbl_mant_dig()) # 64 bit double significand
```

I'm not sure if people would actually be trying to use SWIG on a platform with non-IEEE floating point though.

@wsfulton Thoughts?

Reply:
I sadly still have no better option than hardcoding this constants in the generated C code. And yes, this is hardcoded for IEEE floats only. Do you want me to add some macros to check if the predefined macros in float.h matches our expections and #error else?

However I skipped your suggestion for generating `llmaxd`. This is within the test, and for the cast_mode feature, the test is currently more a documentation of the behavior than checking for the designed behavior. I fear that importing the size of floats from C makes this just more unreadable.

Reply:
Agreed that hard coding is not ideal, but given it is not easily calculated portably, I've put in a comment on how to calculate it as the next best thing. See eac27f7d3ecfe0c120f202eb1a1da081060884f6.